docs/source/runtime_troubles.rst

```diff
@@ -0,0 +1,311 @@
+runtime. This is most useful for resolving :ref:`import cycles <import-cycles>`.
+
+.. note::
+
+   Python 3.5.1 and below don't have :py:data:`~typing.TYPE_CHECKING`. An
+   alternative is to define a constant named ``MYPY`` that has the value
+   ``False`` at runtime. Mypy considers it to be ``True`` when type checking.
+
+Class name forward references
+-----------------------------
+
+Python does not allow references to a class object before the class is
+defined (aka forward reference). Thus this code does not work as expected:
+
+.. code-block:: python
+
+   def f(x: A) -> None: ...  # NameError: name 'A' is not defined
+   class A: ...
+
+Starting from Python 3.7, you can add the special import ``from __future__ import annotations``,
+which makes the use of string literals in annotations unnecessary:
+
+.. code-block:: python
+
+   from __future__ import annotations
+
+   def f(x: A) -> None: ...  # OK
+   class A: ...
+
+For Python 3.6 and below, you can enter the type as a string literal or type comment:
+
+.. code-block:: python
+
+   def f(x: 'A') -> None: ...  # OK
+
+   # Also OK
+   def g(x):  # type: (A) -> None
+       ...
+
+   class A: ...
+
+Of course, instead of using a string literal type or special import, you could move the
+function definition after the class definition. This is not always
+desirable or even possible, though.
+
+.. _import-cycles:
+
+Import cycles
+-------------
+
+An import cycle occurs where module A imports module B and module B
+imports module A (perhaps indirectly, e.g. ``A -> B -> C -> A``).
+Sometimes in order to add type annotations you have to add extra
+imports to a module and those imports cause cycles that didn't exist
+before. This can lead to errors at runtime like:
+
+.. code-block::
+
+   ImportError: cannot import name 'b' from partially initialized module 'A' (most likely due to a circular import)
+
+If those cycles do become a problem when running your program, there's a trick:
+if the import is only needed for type annotations and you're using a) the
+:ref:`future annotations import<future-annotations>`, or b) string literals or type
+comments for the relevant annotations, you can write the imports inside ``if
+TYPE_CHECKING:`` so that they are not executed at runtime. Example:
+
+File ``foo.py``:
+
+.. code-block:: python
+
+   from typing import List, TYPE_CHECKING
+
+   if TYPE_CHECKING:
+       import bar
+
+   def listify(arg: 'bar.BarClass') -> 'List[bar.BarClass]':
+       return [arg]
+
+File ``bar.py``:
+
+.. code-block:: python
+
+   from typing import List
+   from foo import listify
+
+   class BarClass:
+       def listifyme(self) -> 'List[BarClass]':
+           return listify(self)
+
+.. _not-generic-runtime:
+
+Using classes that are generic in stubs but not at runtime
+----------------------------------------------------------
+
+Some classes are declared as generic in stubs, but not at runtime.
```

Comment:
Maybe add a short note explaining the term "generic class" or add a link to relevant section (we don't do this consistently, but I think that it would be useful to have more cross references like these).